airflow/models/dagwarning.py

```diff
@@ -0,0 +1,93 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+from enum import Enum
+
+from sqlalchemy import Column, ForeignKeyConstraint, Integer, String, Text, false
+
+from airflow.models.base import ID_LEN, Base
+from airflow.utils import timezone
+from airflow.utils.session import NEW_SESSION, provide_session
+from airflow.utils.sqlalchemy import UtcDateTime
+
+
+class DagWarning(Base):
+    """
+    A table to store DAG warnings.
+
+    DAG warnings are problems that don't rise to the level of failing the DAG parse
+    but which users should nonetheless be warned about.  These warnings are recorded
+    when parsing DAG and displayed on the Webserver in a flash message.
+    """
+
+    id = Column(
+        Integer, autoincrement=True, server_default='id'
+    )  # this default just signals to SQLA to defer to server
```

Comment:
Do we need the default? I think `autoincrement` will add the needed default. Also, this table has primary key on `dag_id`, `warning_type` but the migration doesn't have these columns as the primary key. Both should have the same configuration. 
Only the ID field is a primary key on the migration file

Reply:
basically, if we want an integer surrogate key we have to make it actually the primary key and use a unique index for the logical key. but then to make sqlalchemy work properly, e.g. with function merge , then sqlalchemy has to think that the logical key is the primary key (even though it’s really not).  finally, the server_default value is required otherwise you’ll get a null constraint violation cus it will try to populate the record with a null value.

the alternative is, (1) no surrogate key, or (2) yes surrogate key but bad sqlalchemy behavior (e.g. merge will be unusable).  in this particular case, having a surrogate key doesn’t really make a difference. BUT i think it’s a “best” practice to have a surrogate key in general (for normalization concerns), so i think we should probably figure out an acceptable approach for it somehow or another.

Reply:
I think that if we use `merge` for updating items and `add` for adding new items we won't have issues with the surrogate key. We can still use `merge` for new items, we just need to ensure that the item is not already in `session` (see https://www.bookstack.cn/read/sqlalchemy-1.3/999f5d55791da984.md). Even the proposed solution will have an issue with `merge` if the item already exists in the session.

Another thing is, would this table confuse other SQLAlchemy users? It feels like a hack to have an `id` field with a `server_default` value and an `autoincrement` set. If we want to use it, then we might need to make a note and also add a link for reference purposes why we need it.

Also, it feels like we will also have issues reconciling the ORM and DB for this case.

Reply:
The whole point of merge is you don't know whether the object is there.  And if it's a composite primary key, it's a pain to build out the query to check for its existence.  That's what's nice about merge.

In any case, we don't really need the surrogate key here so I'll just chop it.  But this will likely come up again at some point.  And it seems this pattern, of using the logical key in ORM while having the surrogate key as the physical key is a good approach for this type of thing.